v1/remote/image_test.go

```diff
@@ -0,0 +1,292 @@
+	}
+	return c
+}
+
+func randomImage(t *testing.T) v1.Image {
+	rnd, err := random.Image(1024, 1)
+	if err != nil {
+		t.Fatalf("random.Image() = %v", err)
+	}
+	return rnd
+}
+
+func newReference(host, repo, ref string) (name.Reference, error) {
+	tag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", host, repo, ref), name.WeakValidation)
+	if err == nil {
+		return tag, nil
+	}
+	return name.NewDigest(fmt.Sprintf("%s/%s@%s", host, repo, ref), name.WeakValidation)
+}
+
+// TODO(jonjohnsonjr): Make this real.
+func TestMediaType(t *testing.T) {
+	img := remoteImage{}
+	got, err := img.MediaType()
+	if err != nil {
+		t.Fatalf("MediaType() = %v", err)
+	}
+	want := types.DockerManifestSchema2
+	if got != want {
+		t.Errorf("MediaType() = %v, want %v", got, want)
+	}
+}
+
+func TestRawManifestDigests(t *testing.T) {
+	img := randomImage(t)
+	expectedRepo := "foo/bar"
+
+	cases := []struct {
+		ref           string
+		responseBody  []byte
+		contentDigest string
+		wantErr       bool
+	}{{
+		// Normal pull, by tag.
+		ref:           "latest",
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       false,
+	}, {
+		// Normal pull, by digest.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       false,
+	}, {
+		// Right content-digest, wrong body, by tag.
+		ref:           "latest",
+		responseBody:  []byte("not even json"),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       true,
+	}, {
+		// Right content-digest, wrong body, by digest.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  []byte("not even json"),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       true,
+	}, {
+		// Right body, wrong content-digest, by tag.
+		ref:           "latest",
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: bogusDigest,
+		wantErr:       true,
+	}, {
+		// Right body, wrong content-digest, by digest.
+		// NB: This succeeds! We don't care what the registry thinks.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: bogusDigest,
+		wantErr:       false,
+	}, {
+		// Nothing matches anything.
+		ref:           "latest",
+		responseBody:  []byte("everything is wrong with this"),
+		contentDigest: bogusDigest,
+		wantErr:       true,
+	}}
+
+	for _, tc := range cases {
+		manifestPath := fmt.Sprintf("/v2/%s/manifests/%s", expectedRepo, tc.ref)
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			switch r.URL.Path {
+			case manifestPath:
+				if r.Method != http.MethodGet {
+					t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+				}
+
+				w.Header().Set("Docker-Content-Digest", tc.contentDigest)
+				w.Write(tc.responseBody)
+			default:
+				t.Fatalf("Unexpected path: %v", r.URL.Path)
+			}
+		}))
+		defer server.Close()
```

Comment:
`defer` is function scope, so each iteration of the loop is going to start a server that stays up until the entire test is complete.  A workaround for this is to wrap the scope in:
```go
func() {
  // Scope of defer
  ...
}()
```

Reply:
Consider using `t.Run` to solve this? https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks

Reply:
Neat, thanks